src/server/auth/server/api_server.go

```diff
@@ -1416,22 +1368,147 @@ func (a *apiServer) RestoreAuthToken(ctx context.Context, req *auth.RestoreAuthT
+	if err := func() error {
+		// Check whether the token hash already exists - we don't want to replace an existing token
+		if _, err := a.lookupAuthTokenInfo(ctx, req.Token.HashedToken); err == nil {
 			return errors.New("cannot overwrite existing token with same hash")
-		} else if err != nil && !col.IsErrNotFound(err) {
+		} else if !col.IsErrNotFound(err) {
 			return err
 		}
-
-		return tokens.PutTTL(req.Token.HashedToken,
-			req.Token.TokenInfo,
-			ttl)
-	}); err != nil {
+		// insert token
+		if ttl > 0 {
+			return a.upsertAuthToken(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject, ttl)
+		} else {
+			return a.upsertAuthTokenNoTTL(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject)
+		}
+	}(); err != nil {
 		return nil, errors.Wrapf(err, "error restoring auth token")
 	}
+
 	return &auth.RestoreAuthTokenResponse{}, nil
 }
+
+// returns -1 if token is not found, and the TTL otherwise
+func (a *apiServer) lookupAuthTokenTTLSeconds(ctx context.Context, tokenHash string) (int64, error) {
+	var ttl float64
+	if err := a.env.GetDBClient().GetContext(ctx, &ttl,
+		`SELECT ROUND( EXTRACT( EPOCH FROM (expiration - NOW()))) as ttl 
+		FROM auth.auth_tokens WHERE token_hash = $1 AND expiration IS NOT NULL`, tokenHash); err != nil {
+		if err == sql.ErrNoRows {
+			return -1, nil
+		}
+		return 0, err
+	}
+	return int64(ttl), nil
+}
+
+func (a *apiServer) lookupAuthTokenInfo(ctx context.Context, tokenHash string) (*auth.TokenInfo, error) {
+	var tokenInfo auth.TokenInfo
+	if err := a.env.GetDBClient().GetContext(ctx, &tokenInfo, `SELECT subject FROM auth.auth_tokens WHERE token_hash = $1`, tokenHash); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, col.ErrNotFound{Type: "auth_tokens", Key: tokenHash}
+		}
+		return nil, err
+	}
+	return &tokenInfo, nil
+}
+
+func (a *apiServer) listRobotTokens(ctx context.Context) ([]*auth.HashedAuthToken, error) {
+	rows, err := a.env.GetDBClient().QueryxContext(ctx,
+		`SELECT token_hash as tokenHash, subject, ROUND( EXTRACT( EPOCH FROM (expiration - NOW()))) as ttl  
+		FROM auth.auth_tokens 
+		WHERE subject LIKE $1 || '%'`, auth.RobotPrefix)
+	if err != nil {
+		return nil, err
+	}
+	robotTokens := make([]*auth.HashedAuthToken, 0)
+	for rows.Next() {
+		var tokenHash string
+		var subject string
+		var ttl *float64
+		if err = rows.Scan(&tokenHash, &subject, &ttl); err != nil {
+			return nil, errors.Wrapf(err, "error querying token")
+		}
+
+		token := &auth.HashedAuthToken{
+			HashedToken: tokenHash,
+			TokenInfo: &auth.TokenInfo{
+				Subject: subject,
+			},
+		}
+		if ttl != nil {
+			expiration, err := types.TimestampProto(time.Now().Add(time.Second * time.Duration(int64(*ttl))))
+			if err != nil {
+				return nil, err
+			}
+			token.Expiration = expiration
+		}
+		robotTokens = append(robotTokens, token)
+	}
+	return robotTokens, nil
+}
+
+func (a *apiServer) generateAndInsertAuthToken(ctx context.Context, subject string, ttlSeconds int64) (string, error) {
```

Comment:
I should add a comment here to clarify that we interpret an expiration value of NULL as "lives forever". Now we will sometimes have expiration values set in the passed, since we only remove those values in the purgeExpiredTokens() goroutine